README
TV-ProgrammesSchedules-SAB

The  README  is  used  to  introduce the module and provide instructions on how to install the
module, any machine dependencies it may have (for example C compilers and installed libraries)
and any other information that should be provided before the module is installed.

A  README  file is required for CPAN modules since CPAN extracts the README file from a module
distribution  so  that  people  browsing the archive can use it to get an idea of the module's
uses.  It is usually a good idea to provide version information here so that people can decide
whether fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

Alternatively, to install with Module::Build, you can use the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc TV::ProgrammesSchedules::SAB

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=TV-ProgrammesSchedules-SAB

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/TV-ProgrammesSchedules-SAB

    CPAN Ratings
        http://cpanratings.perl.org/d/TV-ProgrammesSchedules-SAB

    Search CPAN
        http://search.cpan.org/dist/TV-ProgrammesSchedules-SAB/

LICENSE AND COPYRIGHT

Copyright (C) 2011 Mohammad S Anwar

This  program  is  distributed  in  the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

See http://dev.perl.org/licenses/ for more information.

Changes
Revision history for TV-ProgrammesSchedules-SAB

0.03    Tue Aug 09 11:22:13 2011
        - Made Readonly v1.03 as minimum requirement.
        - Prepared the distribution using the Module::Signature v0.68.

0.02    Thu May 19 11:12:13 2011
        - Updated text for "LICENSE AND COPYRIGHT".

0.01    Tue May 17 14:29:01 2011
        - First version, released.

Build.PL
use strict;use warnings;
use Module::Build;

my $builder = Module::Build->new(
    sign              => 1,
    module_name       => 'TV::ProgrammesSchedules::SAB',
    license           => 'perl',
    dist_author       => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    dist_version_from => 'lib/TV/ProgrammesSchedules/SAB.pm',
    add_to_cleanup    => [ 'TV-ProgrammesSchedules-SAB-*' ],    
    build_requires    => {
        'Test::More'  => 0,
        'Readonly'    => '1.03',
        'Moose'       => '2.0006',
        'Time::localtime'  => '1.02',
        'HTTP::Request'    => '1.40',
        'LWP::UserAgent'   => '2.33',
        'namespace::clean' => '0.20',
        'HTML::TokeParser::Simple'     => '3.15',
        'MooseX::Params::Validate'     => '0.16',
        'Moose::Util::TypeConstraints' => '2.0006',
    },
);

$builder->create_build_script();

lib/TV/ProgrammesSchedules/SAB.pm
package TV::ProgrammesSchedules::SAB;

use Moose;
use MooseX::Params::Validate;
use Moose::Util::TypeConstraints;
use namespace::clean;

use Carp;
use Readonly;
use Data::Dumper;

use HTTP::Request;
use LWP::UserAgent;
use Time::localtime;
use HTML::TokeParser::Simple;

=head1 NAME

TV::ProgrammesSchedules::SAB - Interface to SAB TV Programmes Schedules.

=head1 VERSION

Version 0.03

=cut

our $VERSION = '0.03';
Readonly my $BASE_URL => 'http://www.sabtv.com/comedy/';

=head1 DESCRIPTION

SAB TV is an Indian  general  entertainment  television channel that is owned  by Multi Screen
Media (P) Ltd & based in Mumbai Maharashtra. It was launched on April 23, 2000 by Sri Adhikari 
Brothers as a general entertainment channel. In March 2005, SAB TV was acquired by SET and was 
transformed into a youth-centric channel.

=cut

has 'browser' => (is => 'rw', isa => 'LWP::UserAgent', default => sub { return LWP::UserAgent->new(agent => 'Mozilla/5.0'); });

=head1 METHODS

=head2 get_listings()

Returns the  programmes  listings for the given day. It expects integer optionally as the only
parameter. If missing then it assumes it is 0, which means today. If passed 1 then it means +1
day from today. If passed -2 then it means -2 day from today. Data would be in XML format.

    use strict; use warnings;
    use TV::ProgrammesSchedules::SAB;
    
    my $sab      = TV::ProgrammesSchedules::SAB->new();
    # SAB TV todays listings.
    my $listings = $sab->get_listings();

=cut

sub get_listings
{
    my $self  = shift;
    my ($day) = pos_validated_list(\@_,
                { isa => 'Int', default => 0 },
                MX_PARAMS_VALIDATE_NO_CACHE => 1);

    my ($url, $request, $response, $content);
    my ($listings, $program, $stream, $data);
    my ($token, $tag, $attr, $attrseq, $rawtxt);
    
    
    $url      = $BASE_URL . 'schedule.php?for=' . $day;
    $request  = HTTP::Request->new(GET => $url);
    $response = $self->{browser}->request($request);
    croak("ERROR: Couldn't fetch programmes schedules [$url][".$response->status_line."]\n")
        unless $response->is_success;
    $content  = $response->content;
    croak("ERROR: Data not found [$url].\n")
        unless $content;

    $stream = HTML::TokeParser::Simple->new(string => $content);
    while ($token = $stream->get_tag('td')) 
    {
        ($tag, $attr, $attrseq, $rawtxt) = @{$token};
        if (ref($attr) && exists($attr->{class}) && ($attr->{class} =~ /\bws\_title\b|\bws\_time\b/))
        {
            $data = $stream->get_token();
            ($tag, $attr, $attrseq, $rawtxt) = @{$data};
            if ($tag eq 'T')
            {
                if ($attr =~ /\d{2}\:\d{2}/)
                {
                    push @{$listings}, $program if defined $program;
                    $program = {};
                    $program->{time} = $attr;
                }
                else
                {
                    $program->{title} = $attr;
                }
            }
            elsif ($tag eq 'S')
            {
                $token = $stream->get_token();
                $program->{title} = $token->[1];
                $program->{url}   = $BASE_URL . '/'. $data->[2]->{href};
            }    
        }
    }
    
    warn("WARN: No schedule information found.\n") && return
        unless defined $listings;
        
    $self->{listings} = _toXML($listings);
    return $self->{listings};
}

sub _toXML
{
    my $data = shift;
    my $xml  = qq {<?xml version="1.0" encoding="UTF-8"?>\n};
    $xml.= qq {<programmes>\n};
    foreach (@{$data})
    {
        $xml .= qq {\t<programme>\n};
        $xml .= qq {\t\t<time> $_->{time} </time>\n};
        $xml .= qq {\t\t<title> $_->{title} </title>\n};
        $xml .= qq {\t\t<url> $_->{href} </url>\n} if exists($_->{href});
        $xml .= qq {\t</programme>\n};
    }
    $xml.= qq {</programmes>};
    return $xml;
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-tv-programmesschedules-sab at rt.cpan.org> 
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=TV-ProgrammesSchedules-SAB>.
I will be notified and then you'll automatically be notified of progress on your bug as I make
changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc TV::ProgrammesSchedules::SAB

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=TV-ProgrammesSchedules-SAB>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/TV-ProgrammesSchedules-SAB>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/TV-ProgrammesSchedules-SAB>

=item * Search CPAN

L<http://search.cpan.org/dist/TV-ProgrammesSchedules-SAB/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Mohammad S Anwar.

This  program  is  free  software; you can redistribute it and/or modify it under the terms of
either:  the  GNU  General Public License as published by the Free Software Foundation; or the
Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 DISCLAIMER

This  program  is  distributed in the hope that it will be useful,  but  WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut

__PACKAGE__->meta->make_immutable;
no Moose; # Keywords are removed from the TV::ProgrammesSchedules::SAB package

1; # End of TV::ProgrammesSchedules::SAB

Makefile.PL
use strict;use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME          => 'TV::ProgrammesSchedules::SAB',
    AUTHOR        => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM  => 'lib/TV/ProgrammesSchedules/SAB.pm',
    ABSTRACT_FROM => 'lib/TV/ProgrammesSchedules/SAB.pm',
    dist          => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean         => { FILES => 'TV-ProgrammesSchedules-SAB-*' },
    PREREQ_PM     => {
        'Test::More'  => 0,
        'Readonly'    => '1.03',
        'Moose'       => '2.0006',
        'Time::localtime'  => '1.02',
        'HTTP::Request'    => '1.40',
        'LWP::UserAgent'   => '2.33',
        'namespace::clean' => '0.20',
        'HTML::TokeParser::Simple'     => '3.15',
        'MooseX::Params::Validate'     => '0.16',
        'Moose::Util::TypeConstraints' => '2.0006',
    },
    ($ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
);